[TiangSoonYong] iP

TiangSoonYong opened the pull request.

Commit:
Rename, Greet, Exit

Renamed Duke to Doraemon
Implemented initial skeletal version
Removed traces of Duke from source code

Commit:
Echo

Improved skeletal version by adding Echo capability
Added Scanner library
Created printOut method for printing output

Commit:
Add, List

Renamed printOut to echo
Implemented basic if-else interface to respond to commands
Added List.java to create List class
- Created addTask method
- Created printTasks method
- Fixed to maximum of 100 tasks in the list

Commit:
Mark as Done

Added Task class
Implemented marking and unmarking methods
Renamed to description from task
Note: Can add sanity check

Commit:
Follow the Coding Standard

Reformatted the code
Renamed taskNumber to taskIndex, to be clearer
Realigned certain prints

Commit:
ToDos, Events, Deadlines

Renamed List.java to TaskManager for code clarity
Combined marking methods into single markDone method
Added ToDo, Deadline, Event classes inherited from Task class
Developed addTask to add every type of tasks
Added enum TaskType to handle different types of Task available
Prepared Formatter and Parser classes to handle inputs and outputs in future development

ShengBin-101 commented on src/main/java/Doraemon.java:
```diff
@@ -0,0 +1,74 @@
+
+
+public class Doraemon {
+    // Will go to Formatter class in future development
```
I like how you give comments to remind yourself and inform others of future development, this helps others understand your code design. I noticed this is other parts of your codes, good job!

ShengBin-101 commented on src/main/java/Doraemon.java:
```diff
@@ -0,0 +1,74 @@
+
+    private static final String COMMAND_TODO = "todo";
+    private static final String COMMAND_DEADLINE = "deadline";
+    private static final String COMMAND_EVENT = "event";
```
I like how you made use of constants to define the commands and common messages that are not meant to be mutable. The naming of the constants adhere to coding standard and is clear in telling others the use of the constant.

ShengBin-101 commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,77 @@
+        description = commandArgs.substring(0, indexOfFirstPrefix).trim();
+        if (indexOfFromPrefix < indexOfToPrefix) { // Description - From - To
+            from = removePrefixSign(commandArgs.substring(indexOfFromPrefix, indexOfToPrefix), DATA_PREFIX_FROM).trim();
+            to = removePrefixSign(commandArgs.substring(indexOfToPrefix, commandArgs.length()), DATA_PREFIX_TO).trim();
```
I like that your function clearly describes its own purpose.

ShengBin-101 commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,77 @@
+public class TaskManager {
```
Pascal Casing for class names, adheres to coding standard.

ShengBin-101 commented on src/main/java/TaskType.java:
```diff
@@ -0,0 +1,3 @@
+public enum TaskType{
+    TODO, DEADLINE, EVENT
```
Enums that represent specific task types, useful for further addition of other task types in the future. Adheres to coding standard.

ShengBin-101 reviewed (commented):
Overall, great work with adhering to the coding standards specified. 

- I like how your code includes comments to tell others your future development which helps to convey your code design. 
- Class/Variable/Enum naming follows the coding standard. 
- Functions are well named to convey the use of the function in one glance.

Phua-Lock-Hian commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,22 @@
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public void setDone(boolean done) {
```
the function parameter being done instead of isDone might be misleading, maybe just align with what is existing and use isDone.

TiangSoonYong replied:
Thank you for spotting this error, I shall get it rectified 😄

Phua-Lock-Hian commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,77 @@
+
+        description = commandArgs.substring(0, indexOfFirstPrefix).trim();
+        if (indexOfFromPrefix < indexOfToPrefix) { // Description - From - To
+            from = removePrefixSign(commandArgs.substring(indexOfFromPrefix, indexOfToPrefix), DATA_PREFIX_FROM).trim();
+            to = removePrefixSign(commandArgs.substring(indexOfToPrefix, commandArgs.length()), DATA_PREFIX_TO).trim();
+
+        } else { // Description - To - From
+            from = removePrefixSign(commandArgs.substring(indexOfFromPrefix, commandArgs.length()), DATA_PREFIX_FROM).trim();
+            to = removePrefixSign(commandArgs.substring(indexOfToPrefix, indexOfFromPrefix), DATA_PREFIX_TO).trim();
```
might want to split some lines here, it's getting a bit long and difficult to read.

TiangSoonYong replied:
Agreed! I will create newlines for better code clarity.

Phua-Lock-Hian commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,77 @@
+
+    public String markDone(int taskIndex, boolean done) {
+        tasks[taskIndex].setDone(done);
+        return "Nice! I've marked this task as " + (done ? "done" : "not done yet") + ":\n\t\t " + tasks[taskIndex];
```
the code may be more readable if you move the if-else expression outside before printing.

TiangSoonYong replied:
Agreed, it made the code too long as well!

Phua-Lock-Hian commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,77 @@
+        return message;
+    }
+
+    public String markDone(int taskIndex, boolean done) {
```
this seems like we can only mark tasks as done but the required functionality is to mark or unmark. the functionality looks fine so perhaps just a name change would be good to minimize possible confusion, e.g. setIsDone() instead of markDone().

Phua-Lock-Hian commented on src/main/java/Doraemon.java:
```diff
@@ -0,0 +1,74 @@
+    private static final Scanner SCANNER = new Scanner(System.in);
+
+    public static void main(String[] args) {
+        echo(MESSAGE_HELLO);
```
is echo() still needed or can you just print?

TiangSoonYong replied:
Yes, I use `echo(String... message)` to simplify the process of printing the top and bottom `LINE_DIVIDER` along with displayed message. Additionally, I plan to push all printings to be done by a `Formatter` object, so that my main() will be free of `System.out.println()`.

Phua-Lock-Hian reviewed (commented):
Looks good overall. I like the effort in documenting with comments and noting of future development plans. Personally I feel that there are too many constants used that I have to keep referring back to.

TiangSoonYong reviewed (commented):
Thanks you for the comments! I will rectify the issues and push the updated commit. All the best for your iP too!

Commit:
Resolving PR Comments

Fixed minor formatting of the output in Doraemon.java
Adjusted code line neatness in TaskManager.addEvent
Renamed argument in TaskManager, done to isDone, to follow coding standard
Renamed markDone method to TaskManager.setIsDone, for code clarity
Seperated the if-else statement in TaskManager.setIsDone, for code clarity

Commit:
Automated Text UI Testing

Edited input and runtest files
Copied output from ACTUAL to EXPECTED
Extracted invalid message into a constant

zavsky commented on src/main/java/Doraemon.java:
```diff
@@ -0,0 +1,74 @@
+    private static final String MESSAGE_INVALID = "Invalid Command";
+
+    private static final TaskManager TASK_MANAGER = new TaskManager();
+    private static final Scanner SCANNER = new Scanner(System.in);
```
Should `TASK_MANAGER` and `SCANNER` be camelCase as they are not constants but rather object names?

TiangSoonYong replied:
Both objects were actually declared as a constant with `static final`, thus the naming convention

zavsky commented on src/main/java/Doraemon.java:
```diff
@@ -0,0 +1,74 @@
+            return TASK_MANAGER.addTask(commandArgs, TaskType.EVENT);
+        default:
+            // Invalid command should throw an exception but assume all valid for now
+            return MESSAGE_INVALID;
```
Should the cases each have a `break` statement or a `// Fallthrough` comment here?

TiangSoonYong replied:
I am unsure whether a `break` statement is required as each cases ends with a `return`

zavsky commented on src/main/java/Doraemon.java:
```diff
@@ -0,0 +1,74 @@
+
+    private static String[] splitCommandWordAndArgs(String rawUserInput) {
+        final String[] split = rawUserInput.trim().split("\\s+", 2);
+        return split.length == 2 ? split : new String[]{split[0], ""}; // else case: no parameters
```
Maybe consider splitting this line up as it could be confusing to read?

TiangSoonYong replied:
Agreed! I will change it to individual if-else expression

zavsky commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    private void addToDo(String commandArgs) {
+        String description = commandArgs;
```
Is this necessary? Or is this in preparation for checking if the input is valid?

TiangSoonYong replied:
Actually, I left it as such to follow the same abstraction level. But it seems like I can use it for throwing exception too. Thank you for pointing that out!

zavsky commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,86 @@
+    private void addDeadline(String commandArgs) {
+        final int indexOfByPrefix = commandArgs.indexOf(DATA_PREFIX_BY);
+        String description = commandArgs.substring(0, indexOfByPrefix).trim();
+        String by = removePrefixSign(commandArgs.substring(indexOfByPrefix, commandArgs.length()), DATA_PREFIX_BY).trim();
```
Is this necessary or could the prefix be skipped completely by using `indexOfByPrefix + 1`?

TiangSoonYong replied:
Personally, using the `removePrefixSign` method seems easier for me as it is able to handle any prefixes and I do not need to consider the number of chars to skip. Perhaps another method would be `indexOfByPrefix + DATA_PREFIX_BY.length()`

zavsky commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,86 @@
+        String to;
+
+        description = commandArgs.substring(0, indexOfFirstPrefix).trim();
+        if (indexOfFromPrefix < indexOfToPrefix) { // Description - From - To
+            from = removePrefixSign(
+                    commandArgs.substring(indexOfFromPrefix, indexOfToPrefix),
+                    DATA_PREFIX_FROM).trim();
+            to = removePrefixSign(commandArgs.substring(indexOfToPrefix, commandArgs.length()),
+                    DATA_PREFIX_TO).trim();
```
I like your implementation to check the order of the user commands given to make sure that the program takes in the correct parameters

zavsky commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,86 @@
+
+    public String setIsDone(int taskIndex, boolean isDone) {
+        tasks[taskIndex].setDone(isDone);
+        if (isDone) {
```
I like that your code follows the conventions very well

zavsky reviewed (commented):
Your code is very readable and overall a very good job! Not sure if the `taskCount` updates properly though as it does not seem to increment after task creation. Just a slight nit

TiangSoonYong reviewed (commented):
Thank you for your comments once again! I will make any relevant changes. All the best!

Commit:
Handle Errors

Added various application-specific exceptions as well as its handler
Extracted magic literal, max tasks size, and declared as a constant
Separated if-else expression in splitCommandAndArgs for code clarity

Commit:
Adding help command

Added "help" command and its printing
Reformatted the code

Commit:
Updated input text files to include exception testings
Updated actual text files according to new input

Commit:
Merge branch 'branch-Level-5'

Commit:
Organize into Packages

Separated classes into different packages, exceptions and task packages
Updated runtest.bat to work with new directory

Commit:
Merge branch 'branch-A-Packages'

TVageesan commented on src/main/java/doraemon/Doraemon.java:
```diff
@@ -0,0 +1,124 @@
+    private static final String LINE_DIVIDER = "__________________________________________________";
+
+    // Will go to Parser class in future development
+    private static final String LS = "\n\t "; // Line Separator
```
Name this LINE_SEPERATOR. In general, avoid abbreviations. This [video](https://youtu.be/-J3wNP6u5YU?t=69) has a really good explanation at to why (just the 1:00-2:00 part)

TiangSoonYong replied:
That's a really good point that I did not consider!

TVageesan commented on src/main/java/doraemon/task/TaskManager.java:
```diff
@@ -0,0 +1,150 @@
+            default:
+                // Throw exception
+            }
+        } catch (NoDescriptionException e) {
+            return "The description is empty!";
+        } catch (NoByPrefixException e) {
+            return "By prefix [/by] is missing!";
+        } catch (NoByStringException e) {
+            return "The deadline is empty!";
+        } catch (NoFromPrefixException e) {
+            return "From prefix [/from] is missing!";
+        } catch (NoFromStringException e) {
+            return "The start date is missing!";
+        } catch (NoToPrefixException e) {
+            return "To prefix [/to] is missing!";
+        } catch (NoToStringException e) {
+            return "The end date is missing!";
+        }
```
Good idea to use Exceptions to simplify error handling! You could make this code even simpler by making the exceptions all inherit from a common exception (maybe MissingArgumentException), then just have a generic `catch (MissingArgumentException e) { return e.message }`.

TiangSoonYong replied:
Oh wow, that is a great idea. That way, I won't have to keep adding on more catches for further developments!

TVageesan commented on src/main/java/doraemon/task/TaskManager.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    private void addEvent(String commandArgs)
+            throws NoDescriptionException, NoFromPrefixException, NoFromStringException,
+            NoToPrefixException, NoToStringException {
```
Following from above, having the generic exception would also simplify declarations like this to `throws MissingArgumentException`

TiangSoonYong replied:
I see, thank you for this suggestion!

TVageesan commented on src/main/java/doraemon/task/TaskManager.java:
```diff
@@ -0,0 +1,150 @@
+
+    public String setIsDone(int taskIndex, boolean isDone) {
+        try {
+            tasks[taskIndex].setDone(isDone);
```
Instead of directly manipulating the tasks array, `tasks[taskIndex]` consider putting it into a suitable variable first. Here, it could be `selectedTask = tasks[taskIndex]; selectedTask.setDone(isDone)`. this makes it easier to read and understand the code as the reader doesnt have to understand what tasks array is or what data it holds.

TiangSoonYong replied:
I like your suggestion!

TVageesan reviewed (commented):
Overall, really well done. Honestly couldn't find much wrong readability wise with your code, and your future implementations of Exceptions/Formatter/Parser are solid design decisions. Left some nitpicks and suggestions to further improve.

Commit:
Use Java Collections classes

Changed Tasks array into a Tasks ArrayList
Split help messages into individual usage info for each commands

Commit:
Delete

Add support for deleting tasks from the list
Updated automated testing files to reflect delete process

Commit:
Save

Added support for saving tasks into a tasks.txt in a relative path from project
Added skeleton for reading tasks from /data/tasks.txt, still require parser

Commit:
Read

Added support for reading from text file inclusive of parser

Commit:
Merge branch 'branch-Level-6'

Commit:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/doraemon/task/TaskManager.java

Commit:
Exception Handling

Added exception handler for incorrect format during file reading
Added usage info for delete, save and clear
Reformatted code to use ArrayList methods
Updated automated testing

Commit:
Create a JAR File

TiangSoonYong reviewed (commented):
Thank you for the comments and suggestions. They were great and I will implement them in the next commit!

Commit:
AddTaskException

Created new abstract class AddTaskException which all relevant exceptions inherits from and must implement a getErrorMessage method
Simplified the codes surrounding exceptions
Created local variable selectedTask for better code clarity

Commit:
Added  Storage Class

Commit:
Added Formatter class

Commit:
Added Parser Class

Shifted printed Strings to Formatter Class

Commit:
Added Command class and subclasses

Shifted getCommand to Parser class

Commit:
Extended Exception Handlers

Created a general DoraemonException
Added InvalidTaskNumberException to handle NumberFormatException

Commit:
Reformatted printings

Made the program echo the command given as it is useful to see in automated testing
Shifted TaskManager class to the Doraemon package

Commit:
Merge branch 'branch-A-MoreOOP'

Commit:
Find

Give users a way to find a task by searching for a keyword in the task description.
Added exception when keyword is empty

Commit:
Minor refactoring

For code clarity

Commit:
Added usage info for find command

Commit:
Dates and Times

Store dates and times as LocalDateTimes instead of String

Commit:
Added support to see upcoming tasks

Created DateTimeTask to support Upcoming command

Commit:
Check whether dateTimeTask is overdue

Displays overdue when tasks are listed

Commit:
Added usage info for upcoming command

Commit:
JavaDoc

Add JavaDoc comments
Added NoTaskTypeException to handle default case

TiangSoonYong committed:
Merge pull request #1 from TiangSoonYong/branch-A-JavaDoc

JavaDoc

TiangSoonYong committed:
Merge pull request #2 from TiangSoonYong/branch-Level-8

Branch level 8

TiangSoonYong committed:
Merge branch 'master' into branch-Level-9

TiangSoonYong committed:
Merge pull request #3 from TiangSoonYong/branch-Level-9

Branch level 9

Commit:
Reformatting

Shifted usage info into individual commands
Edited automated testing
Note: Unsure why ? appears in text testing

Commit:
Exception Handling for invalid duration

Commit:
Updated java doc for earlier commit

Commit:
Added javadoc for future development of event and deadline task creation

Commit:
Add a User Guide

Added complete user guide
Reformatted part of codes

Commit:
Merge branch 'branch-A-UserGuide'

Commit:
Resolved error when using LocalDateTime

Used a custom pattern instead of inbuilt formats
Changed all relevant printings as well as UG
Updated text testing

Commit:
Updated InvalidFormatException error message

Format was updated to the current DateTime format
